src/zepben/evolve/services/network/tracing/networktrace/network_trace.py

```diff
@@ -124,28 +133,67 @@ def add_start_item(self, start: Union[Terminal, ConductingEquipment], data: T= N
         :param data: The data associated with the start step.
         :param phases: Phases to trace; `None` to ignore phases.
         """
-        if isinstance(start, Terminal):
-            start_path = NetworkTraceStep.Path(start, start, self.start_nominal_phase_path(phases))
-            super().add_start_item(NetworkTraceStep(start_path, 0, 0, data))
-            return self
+        raise Exception('INTERNAL ERROR:: unexpected add_start_item params')
 
-        if issubclass(start.__class__, ConductingEquipment) or isinstance(start, ConductingEquipment):
-            for it in start.terminals:
-                self.add_start_item(it, data, phases)
-            return self
+    @add_start_item.register
+    def _(self, start: ConductingEquipment, data=None, phases=None):
+        # We don't have a special case for Clamp here because we say if you start from the whole Clamp rather than its terminal specifically,
+        # we want to trace externally from it and traverse its segment.
+        for it in start.terminals:
+            self._add_start_item(it, data, phases, None)
 
-        super().add_start_item(start)
         return self
 
+    @add_start_item.register
+    def _(self, start: Terminal, data=None, phases=None):
+        # We have a special case when starting specifically on a clamp terminal that we mark it as having traversed the segment such that it
+        # will only trace externally from the clamp terminal. This behaves differently to when the whole Clamp is added as a start item.
+        traversed_ac_line_segment = None
+        if isinstance(start.conducting_equipment, Clamp):
+            traversed_ac_line_segment = start.conducting_equipment.ac_line_segment
+        self._add_start_item(start, data, phases, traversed_ac_line_segment)
+        return self
+
+    @add_start_item.register
+    def _(self, start: AcLineSegment, data=None, phases=None):
```

Comment:
question: Does the `@singledispatchmethod` guarantee to call this with all `AcLineSement` and never the `ConductingEquipment` version above?

I assume because the variable type is just a hint and isn't in the run code, it does something different when it is called, but what makes sure it doesn't match the `ConductingEquipment` version above?

Reply:
from all of my testing, yes, and also from the docs, as i understand it, it dispatches based on mro, so the best match to the classes type is used.